drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+		(*buf)->len = (*buf)->size;
+	}
+	return WHD_SUCCESS;
+}
+
+static int airoc_mgmt_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct airoc_wifi_data *data = dev->data;
+	cy_rslt_t ret = CY_RSLT_SUCCESS;
+	size_t pkt_len = net_pkt_get_len(pkt);
+	struct net_buf *buf = NULL;
+
+	/* Allocate Network Buffer from pool with Packet Length + Data Header */
+	buf = net_buf_alloc_len(&airoc_pool, pkt_len + sizeof(data_header_t), K_NO_WAIT);
+	if (buf == NULL) {
+		return -EIO;
+	}
+
+	/* Reserve the buffer Headroom for WHD Data header */
+	net_buf_reserve(buf, sizeof(data_header_t));
+
+	/* Read the packet payload */
+	if (net_pkt_read(pkt, buf->data, pkt_len) < 0) {
+		LOG_ERR("net_pkt_read failed");
+		return -EIO;
+	}
+
+	/* Call WHD API to send out the Packet */
+	ret = whd_network_send_ethernet_data(airoc_if, (void *)buf);
+	if (ret != CY_RSLT_SUCCESS) {
+		LOG_ERR("whd_network_send_ethernet_data failed");
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+		data->stats.errors.tx++;
+#endif
+		return -EIO;
+	}
+
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+	data->stats.bytes.sent += pkt_len;
+	data->stats.pkts.tx++;
+#endif
+
+	return 0;
+}
+
+void cy_network_process_ethernet_data(whd_interface_t interface, whd_buffer_t buffer)
+{
+	struct net_pkt *pkt;
+	uint8_t *data = whd_buffer_get_current_piece_data_pointer(interface->whd_driver, buffer);
+	uint32_t len = whd_buffer_get_current_piece_size(interface->whd_driver, buffer);
+	bool net_pkt_unref_flag = false;
+
+	if ((interface->role == WHD_STA_ROLE) &&
+	    (net_if_flag_is_set(airoc_wifi_iface, NET_IF_UP))) {
+		if (airoc_wifi_iface == NULL) {
+			LOG_ERR("network interface unavailable");
+			cy_buffer_release(buffer, WHD_NETWORK_RX);
+			return;
+		}
+
+		pkt = net_pkt_rx_alloc_with_buffer(airoc_wifi_iface, len, AF_UNSPEC, 0, K_NO_WAIT);
+
+		if (pkt != NULL) {
+			if (net_pkt_write(pkt, data, len) < 0) {
+				LOG_ERR("Failed to write pkt");
+				net_pkt_unref_flag = true;
+			}
+
+			if ((net_pkt_unref_flag) || (net_recv_data(airoc_wifi_iface, pkt) < 0)) {
+				LOG_ERR("Failed to push received data");
+				net_pkt_unref_flag = true;
+			}
+		} else {
+			LOG_ERR("Failed to get net buffer");
+		}
+	}
+	/* Release a packet buffer */
+	cy_buffer_release(buffer, WHD_NETWORK_RX);
+
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+	airoc_wifi_data.stats.bytes.received += len;
+	airoc_wifi_data.stats.pkts.rx++;
+#endif
+
+	if (net_pkt_unref_flag) {
+		net_pkt_unref(pkt);
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+		airoc_wifi_data.stats.errors.rx++;
+#endif
+	}
+}
+
+static void *link_events_handler(whd_interface_t ifp, const whd_event_header_t *event_header,
+				 const uint8_t *event_data, void *handler_user_data)
+{
+	ARG_UNUSED(ifp);
+	ARG_UNUSED(event_data);
+	ARG_UNUSED(handler_user_data);
+
+	(void)k_msgq_put(&airoc_wifi_msgq, event_header, K_FOREVER);
+	return NULL;
+}
+
+static void airoc_event_task(void)
+{
+	whd_event_header_t event_header;
+
+	while (1) {
```

Comment:
Given this is a forever loop, we should have a sleep/yield/whatever call to avoid starving lower priority tasks.

Reply:
I guess `k_msgq_get` blocks if there is no event and that should yield.